### PR TITLE
A fix for Mac OSX process start_time

### PR DESCRIPTION
**Pull request by hughneale**
This hopefully (works for me at least) fixes the issue presented in ticket #3495. Needs to be tested on older OSX builds.

### CODE REVIEW

File: osquery/tables/system/darwin/processes.cpp [hunk 1]
```diff
@@ -310,9 +312,19 @@ QueryData genProcesses(QueryContext& context) {
...
       // Convert the time in CPU ticks since boot to seconds.
       // This is relative to time not-sleeping since boot.
-      r["start_time"] =
-          TEXT((rusage_info_data.ri_proc_start_abstime / START_TIME_RATIO) *
-               time_base.numer / time_base.denom);
+
+      long uptime = tables::getUptime();
+      uint64_t absoluteTime = mach_absolute_time();
+
+      double multiply = (double) time_base.numer / (double) time_base.denom;
```
Comment (obelisk) [inline osquery/tables/system/darwin/processes.cpp]: Prefer auto and static_casts to C-Style casts

File: osquery/tables/system/darwin/processes.cpp [hunk 2]
```diff
@@ -310,9 +312,19 @@ QueryData genProcesses(QueryContext& context) {
       r["system_time"] = TEXT(rusage_info_data.ri_system_time / CPU_TIME_RATIO);
       // Convert the time in CPU ticks since boot to seconds.
       // This is relative to time not-sleeping since boot.
-      r["start_time"] =
-          TEXT((rusage_info_data.ri_proc_start_abstime / START_TIME_RATIO) *
-               time_base.numer / time_base.denom);
+
+      long uptime = tables::getUptime();
```
Comment (obelisk) [inline osquery/tables/system/darwin/processes.cpp]: auto

File: osquery/tables/system/darwin/processes.cpp [hunk 3]
```diff
@@ -310,9 +312,19 @@ QueryData genProcesses(QueryContext& context) {
...
+      uint64_t absoluteTime = mach_absolute_time();
```
Comment (obelisk) [inline osquery/tables/system/darwin/processes.cpp]: auto

File: osquery/tables/system/darwin/processes.cpp [hunk 4]
```diff
@@ -310,9 +312,19 @@ QueryData genProcesses(QueryContext& context) {
...
+
+      double multiply = (double) time_base.numer / (double) time_base.denom;
+      long diff = ((rusage_info_data.ri_proc_start_abstime - absoluteTime));
```
Comment (obelisk) [inline osquery/tables/system/darwin/processes.cpp]: auto

File: osquery/tables/system/darwin/processes.cpp [hunk 5]
```diff
@@ -310,9 +312,19 @@ QueryData genProcesses(QueryContext& context) {
...
+
+      // This is a negative value
+      long seconds_since_launch = (long) ((diff * multiply) / START_TIME_RATIO);
```
Comment (obelisk) [inline osquery/tables/system/darwin/processes.cpp]: auto and static_cast

**Comment by obelisk** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit b2167877b5954e90810930912343dab32ac08bfb (Job results: [2173](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2173/)) failed the code audit or documentation test.

**Comment by hughneale** [comment]
Thanks for your comments. Updated.

**Comment by osqueryer** [comment]
:no_good: The commit 26abb0d44895efcfb69dfd7250f6ebfbdd138d82 (Job results: [2175](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2175/)) failed the code audit or documentation test.

**Comment by obelisk** [comment]
After make format this looks good to me :)

File: osquery/tables/system/darwin/processes.cpp [hunk 6]
```diff
@@ -310,9 +312,23 @@ QueryData genProcesses(QueryContext& context) {
       r["system_time"] = TEXT(rusage_info_data.ri_system_time / CPU_TIME_RATIO);
       // Convert the time in CPU ticks since boot to seconds.
       // This is relative to time not-sleeping since boot.
-      r["start_time"] =
-          TEXT((rusage_info_data.ri_proc_start_abstime / START_TIME_RATIO) *
-               time_base.numer / time_base.denom);
+
+      //
```
Comment (obelisk) [inline osquery/tables/system/darwin/processes.cpp]: ?

File: osquery/tables/system/darwin/processes.cpp [hunk 7]
```diff
@@ -310,9 +312,23 @@ QueryData genProcesses(QueryContext& context) {
...
+      auto diff = static_cast<long>(
+          (rusage_info_data.ri_proc_start_abstime - absoluteTime));
+
+      // This is a negative value
+      auto seconds_since_launch =
+          static_cast<long>((diff * multiply) / START_TIME_RATIO);
+
+      // Get the start_time since the computer started
+      r["start_time"] = TEXT(uptime + seconds_since_launch);
+
```
Comment (obelisk) [inline osquery/tables/system/darwin/processes.cpp]: Nit: nix

File: osquery/tables/system/darwin/processes.cpp (hunk 6 above)
Comment (hughneale) [inline osquery/tables/system/darwin/processes.cpp]: Opps. Sorry. Was trying to figure make format. Wasn't clipping the line length.

File: osquery/tables/system/darwin/processes.cpp (hunk 7 above)
Comment (hughneale) [inline osquery/tables/system/darwin/processes.cpp]: nix?

**Pull request merged**